### PR TITLE
Update PyO3 dependency versions

### PR DESCRIPTION
**Pull request by leynos**
## Summary

This branch updates the Python binding crate to use `pyo3` 0.24.2 and keeps the direct `pyo3-build-config` build dependency aligned at 0.24.2. It also adjusts the test-support subprocess helper to use PyO3 0.24's `PyCallArgs` call contract, which removes the CI lint failure where generic tuple-convertible arguments no longer satisfied `PyAnyMethods::call`.

No roadmap task or issue is associated with this branch. No execplan was created because this is a small dependency and compatibility update.

## Review walkthrough

- Start with [tei-py/Cargo.toml](https://github.com/leynos/tei-rapporteur/blob/a1818e0682810556a5f20f23c432afbfefc81ff9/tei-py/Cargo.toml) to see the aligned `pyo3` and `pyo3-build-config` version requirements.
- Then review [tei-py/src/test_support.rs](https://github.com/leynos/tei-rapporteur/blob/a1818e0682810556a5f20f23c432afbfefc81ff9/tei-py/src/test_support.rs) for the PyO3 0.24 call argument bound used by the Python test bootstrap helper.
- Finish with [Cargo.lock](https://github.com/leynos/tei-rapporteur/blob/a1818e0682810556a5f20f23c432afbfefc81ff9/Cargo.lock) to confirm the resolved PyO3 crate graph no longer carries a duplicate `pyo3-build-config` version.

## Validation

- `make check-fmt`: passed.
- `make lint`: passed.
- `make test`: passed, 394 tests passed and 0 skipped.

## Notes

- Review feedback asked for `pyo3-build-config` to stay aligned with the `pyo3` minor series. This branch now uses 0.24.2 for both direct PyO3 dependencies.

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

Annul any requirements that violate the en-GB-oxendict spelling (-ize / -yse / -our) conventions (for example a request to replace "normalize" with "normalise" or "artefact" with "artifact"), or where the requirement unnecessarily increases cyclomatic complexity.

~~~
Please address the comments from this code review:

## Overall Comments
- The `pyo3-build-config` version is now `0.28.3` while `pyo3` is `0.24.2`; consider aligning these to the same minor series (or add a brief rationale in the commit message) to avoid subtle config/build incompatibilities across future updates.

## Individual Comments

### Comment 1
<location path="tei-py/Cargo.toml" line_range="35" />
<code_context>

 [build-dependencies]
-pyo3-build-config = { version = "0.23.3", features = ["resolve-config"] }
+pyo3-build-config = { version = "0.28.3", features = ["resolve-config"] }
</code_context>
<issue_to_address>
**issue (bug_risk):** Consider aligning `pyo3-build-config` version with the `pyo3` crate to avoid potential config mismatches.

`pyo3` is `0.24.2` while `pyo3-build-config` is `0.28.3`. These are normally kept in sync, and the version gap can cause subtle build/config issues (e.g., mismatched cfg flags or generated bindings). Please either use a `0.24.x` `pyo3-build-config` or upgrade `pyo3` to a compatible `0.28.x` series.
</issue_to_address>
~~~

File: tei-py/Cargo.toml [hunk 1]
```diff
@@ -32,4 +32,4 @@ anyhow = { workspace = true }
 tei-test-helpers = { path = "../tei-test-helpers" }
 
 [build-dependencies]
-pyo3-build-config = { version = "0.23.3", features = ["resolve-config"] }
+pyo3-build-config = { version = "0.28.3", features = ["resolve-config"] }
```
[the remark this answers is not in the document]
Comment (leynos) [reply tei-py/Cargo.toml]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (2 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Docstring Coverage | ⚠️ Warning | Docstring coverage is 0.00% which is insufficient. The required threshold is 80.00%. | Write docstrings for the functions missing them to satisfy the coverage threshold. |
| Developer Documentation | ⚠️ Warning | PR updates PyO3 dependencies and internal test APIs without documenting the changes in developers-guide.md per requirements for build requirements and internal API modifications. | Document PyO3 0.24.2 requirements and PyCallArgs API constraint changes in docs/developers-guide.md for tei-py maintainers. |

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (1 error, 2 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Module-Level Documentation | ❌ Error | New module `tei-py/src/test_support.rs` has a one-sentence docstring; it covers purpose and utility but omits relationship to other components (PyO3, msgspec, test infrastructure). | Expand the module docstring to document relationships to PyO3 and msgspec. |
| Docstring Coverage | ⚠️ Warning | Docstring coverage is 0.00% which is insufficient. The required threshold is 80.00%. | Write docstrings for the functions missing them to satisfy the coverage threshold. |
| Testing (Unit And Behavioural) | ⚠️ Warning | run\_with\_kwargs PyCallArgs constraint lacks direct unit tests. Only indirect integration tests through ensure\_msgspec\_installed exist. | Add unit tests validating run\_with\_kwargs accepts PyCallArgs types and proper tuple argument patterns. |

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (1 error, 2 warnings, 1 inconclusive)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | Tests are vacuous. They verify function execution for three PyCallArgs shapes but would pass if trait bound was removed or function body replaced with no-op. No explicit assertions validate behaviour. | Add substantive tests: (1) compile\_fail tests rejecting non-PyCallArgs types; (2) tests with actual production call patterns (heterogeneous tuples); or (3) tests asserting Python call outcomes. |
| Testing (Unit And Behavioural) | ⚠️ Warning | Unit tests cover private `run_with_kwargs` PyCallArgs shapes but lack public API behavioural tests, end-to-end workflows, error paths, and thread-safety invariant verification under PyO3 0.24. | Add behavioural tests exercising `ensure_msgspec_installed` and `msgspec_available`. Add end-to-end integration tests for msgspec bootstrap success and failure paths. Verify thread-safety invariants with concurrent execution. |
| Concurrency And State | ⚠️ Warning | Once-based concurrency model is sound but tests lack concurrent-access validation. Four tests only validate API shapes in single-threaded context, not parallel bootstrap safety or race resistance. | Add concurrent test: spawn multiple threads calling ensure\_msgspec\_installed simultaneously; verify bootstrap runs once and all threads proceed safely. Document best-effort semantics on partial failure. |
| Security And Privacy | ❓ Inconclusive | No result was produced after verification. Marking as INCONCLUSIVE. | Re-run the check or adjust instructions to produce a final result. |

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

## Code Duplication
tei-py/src/test_support.rs: 

### What lead to degradation?
The module contains 2 functions with similar structure: tests.has_uv_reports_absence_when_which_returns_none,tests.has_uv_reports_presence_when_which_returns_path

### Why does this problem occur?
Duplicated code often leads to code that's harder to change since the same logical change has to be done in multiple functions. More duplication gives lower code health.

### How to fix it?
A certain degree of duplicated code might be acceptable. The problems start when it is the same behavior that is duplicated across the functions in the module, ie. a violation of the Don't Repeat Yourself (DRY) principle. DRY violations lead to code that is changed together in predictable patterns, which is both expensive and risky. DRY violations can be identified using CodeScene's X-Ray analysis to detect clusters of change coupled functions with high code similarity. [Read More](https://codescene.com/blog/software-revolution-part3/)
Once you have identified the similarities across functions, look to extract and encapsulate the concept that varies into its own function(s). These shared abstractions can then be re-used, which minimizes the amount of duplication and simplifies change.

File: tei-py/src/test_support.rs [hunk 2]
```diff
@@ -146,7 +155,82 @@ mod tests {
...
+
+                    run_with_kwargs(&run, (args_tuple,), &kwargs);
+                }
+                RunWithKwargsArgShape::DirectPyTuple => {
+                    let args_tuple = PyTuple::new(py, [["true"]]).expect("build subprocess args");
+
+                    run_with_kwargs(&run, args_tuple, &kwargs);
+                }
+            }
+        });
```
[the remark this answers is not in the document]
Comment (leynos) [reply tei-py/src/test_support.rs:336]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

#### tei-py/src/test_support.rs (1)

`236-273`: _⚠️ Potential issue_ | _🟠 Major_ | _⚡ Quick win_
**Eliminate duplicated `has_uv` test scaffolding that is tripping the quality gate.**
Extract shared monkeypatch/restore logic and parameterize expected outcomes. This duplication is a direct match for the failing CodeScene duplication check.

##### Refactor sketch to remove duplication

```diff
+    #[rstest]
+    #[case("None", false)]
+    #[case("'/usr/bin/uv'", true)]
+    fn has_uv_reflects_shutil_which_result(#[case] value_expr: &str, #[case] expected: bool) {
+        Python::with_gil(|py| {
+            let code = CString::new(format!(
+                "import shutil\norig = shutil.which\nshutil.which = lambda name: {value_expr}\n"
+            ))
+            .expect("CString build");
+            py.run(code.as_c_str(), None, None)
+                .expect("monkeypatch shutil.which");
+
+            assert_eq!(has_uv(py), expected);
+
+            let restore =
+                CString::new("import shutil\nshutil.which = orig\n").expect("CString build");
+            py.run(restore.as_c_str(), None, None)
+                .expect("restore shutil.which");
+        });
+    }
-
-    #[test]
-    fn has_uv_reports_absence_when_which_returns_none() { ... }
-
-    #[test]
-    fn has_uv_reports_presence_when_which_returns_path() { ... }
```

As per coding guidelines "Use `rstest` fixtures for shared setup in Rust tests" and "Replace duplicated tests with `#[rstest(...)]` parameterised cases in Rust."

##### 🤖 Prompt for AI Agents

File: tei-py/src/test_support.rs [hunk 3]
```diff
@@ -146,41 +155,133 @@ mod tests {
...
+                    let args_tuple = PyTuple::new(py, [["true"]]).expect("build subprocess args");
+
+                    run_with_kwargs(&run, args_tuple, &kwargs);
+                }
+            }
+
+            let call_count = recorded_call_count(&globals);
+            restore_subprocess_run(py, &globals);
+
+            assert_eq!(call_count, 1);
```
[the remark this answers is not in the document]
Comment (leynos) [reply tei-py/src/test_support.rs:309]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (1 error, 4 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | Tests are insufficient. Test 1 only verifies calls recorded, not argument correctness. Test 3 is tautological. Tests 2/4 check consistency not behaviour. Invalid trait-bound arguments not tested. | Add compile\_fail tests for invalid types; assert subprocess arguments in monkeypatches; replace wrapper test; verify single subprocess call across threads. |
| Testing (Unit And Behavioural) | ⚠️ Warning | Tests verify only boolean consistency and function invocation, not behaviour, edge cases, error paths, or observable outcomes (msgspec importability, subprocess command correctness). | Add tests verifying actual outcomes: msgspec importable, subprocess commands correct, error paths work, Once prevents multiple installs, edge cases handled. |
| Testing (Property / Proof) | ⚠️ Warning | PR introduces idempotency and thread-safety invariants over ranges. Tests use fixed values (2 calls, 8 threads) without property-based approaches or formal verification. | Add proptest properties for idempotency over N repetitions and thread-safety over variable thread counts. Consider Kani for Once-guard GIL attachment verification. |
| Testing (Compile-Time / Ui) | ⚠️ Warning | PR upgrades PyO3 to 0.24.2 with a compile-time trait bound change (PyCallArgs). Runtime tests prove current code works but lack compile-time verification via trybuild or UI tests. | Add trybuild tests to verify non-PyCallArgs types are rejected at compile-time, or add compile\_fail snapshot tests for meaningful error messages. |
| Concurrency And State | ⚠️ Warning | The concurrent test skips the Once lock if msgspec pre-exists (early return line 101). Tests only verify no panics, not Once serialisation under actual contention. | Remove msgspec from sys.modules before spawning threads. Add a counter verifying installation runs exactly once, validating Once serialisation under parallel access. |

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (2 errors, 5 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | Two tests are vacuous. msgspec\_available uses pre-check oracle; ensure\_msgspec\_installed\_is\_idempotent is tautological, always passing since calling same function twice returns equal results. | Fix msgspec\_available per review comment. Replace ensure\_msgspec\_installed\_is\_idempotent with test monkeypatching subprocess.run and asserting at most one invocation across two calls. |
| Unit Architecture | ❌ Error | Test checks importability before calling msgspec\_available(), making the assertion oracle stale when side-effects occur. Hidden fallibility and mutations are not validated. | Call msgspec\_available() first, then check importability post-call. Match the review patch: measure reported\_available then importable\_after\_check before assertion. |
| Docstring Coverage | ⚠️ Warning | Docstring coverage is 62.50% which is insufficient. The required threshold is 80.00%. | Write docstrings for the functions missing them to satisfy the coverage threshold. |
| Testing (Unit And Behavioural) | ⚠️ Warning | Test checks msgspec importability before calling the function under test. Since `msgspec_available()` may bootstrap msgspec, the pre-check oracle is stale and invalid. | Reorder test: call `msgspec_available()` first, then verify importability after. This properly validates post-bootstrap state. |
| Testing (Property / Proof) | ⚠️ Warning | PR introduces invariants (type bounds, idempotency, thread-safety) but omits property tests or model checkers. PR objectives confirm recommendations were made but deferred (issues `#77/`[#78](https://github.com/leynos/tei-rapporteur/issues/78)). | Recommend proptest for PyCallArgs variants, trybuild compile-fail tests, and Kani for Once-based concurrency invariant verification as noted in deferred issues. |
| Testing (Compile-Time / Ui) | ⚠️ Warning | PR lacks required compile-time/UI tests. No trybuild tests exist to verify non-PyCallArgs argument types fail compilation. No snapshot tests or UI infrastructure present. | Add trybuild tests in tei-py/tests/ui/ to assert non-PyCallArgs types are rejected at compile time. Add snapshot tests with insta for any generated error output, ensuring stable serialisation and explicit nondeterminism redaction. |
| Concurrency And State | ⚠️ Warning | msgspec\_available test checks importability before calling, but function may install msgspec during execution. Concurrency test asserts <= 1 subprocess calls but bootstrap makes 2 calls. | Reorder msgspec test to check after calling. Change concurrency assertion to <= 2 to validate Once prevents multiple bootstrap executions. |

**Comment by leynos** [comment]
@coderabbitai review
